Georef: Accurate street network geometries

eturck opened the pull request:
Hi!

I started this long ago, but I finally took the time to finalize it, fix a few bugs and add some unit tests. I know that you are in the process of changing almost everything on the street network routing algorithm so I hope it's not too late...

There is two commits fixing some bugs : 
- In routing_api_test  an edge added in the way was reversed compare to the one added via boost::add_edge. It was breaking the unit test,
- In osm2ed the geometry of the reversed edge wasn't reversed (since it's not currently used except to get its length there is currently no impact).

Otherwise the main changes are : 
- Export geometries of edges in ed2nav. To reduce memory usage we store the geometry once in the way, and reference it on each type of edge (pedestrian, bike, car),
- Improve projection and distance from both vertex by using the geometry of the edge to find the closest point,
- way::make_multiline use them too,
- Better geojson for street network sections : 
  - Each section use the geometry of the edge,
  - When adding custom projection to the path we split the first and last edge geometries and just keep the part before / after the projected point,
  - Handle the case of two point projected on the same edge too. 

As discussed a few month ago, it's completely optional since it's quite memory intensive. You need to pass an argument to ed2nav otherwise geometries are not exported. On our (small) network : 
- ~15k vj, ~350k stop_times,
- OSM data just on the PTU (https://perspectivedestransports.files.wordpress.com/2013/08/carte-toulouse_rc3a9seau-toulouse.jpg) => 64Mo pbf file.

We have :

|  | lz4 size | RAM used by Kraken at startup |
| --- | --- | --- |
| Without geometries | 44Mo | ~190Mo |
| With geometries | 69Mo | ~240Mo |

And all of this to get : 
![with_geom](https://cloud.githubusercontent.com/assets/5136553/17433473/2a19b082-5b05-11e6-9708-fa3a99ae6d9e.jpg)

Instead of :
![without_geom](https://cloud.githubusercontent.com/assets/5136553/17433484/3a1190ae-5b05-11e6-9521-f70f7e9d4a03.jpg)

Let me know what you think about all of this.

Commit:
georef: Use real geometries for street network edges.

We're doing this for multiple uses :
 - Use real geometries to ensure projection are really on the edge geometry,
 - Use real geometries in the result path geojson,
 - Handle first / last edge and same edge projection by cutting geometry appropriately
 - Use real geometries when searching for nearest edge
 - Compute real distance following geometries for ProjectionData

Commit:
ed: Make export of full street network geometries optional

You need to pass an argument to ed2nav in order to have the full street
network geometries in the lz4, since it's quite memory intensive it's
disabled by default.

Commit:
ed: Add geometries to each edge of the graph.

We're storing them in the way since we're using them in the
make_multiline function. Since the walk / bike / car edge share
the same geometry we're only storing it once.

Commit:
type: add functions to geographical_coord

A function to split a LineString at a certain point and get the start or
end of the geom from this point.
Another one to compute the exact distance between a point and the source
or the target of the edge (using its geometry).

Commit:
Routing: fix an edge in routing_api_test.cpp

The edge added in the way was reversed of the one added to the
graph via boost::add_edge.

Commit:
ed: bugfix on osm2ed. The geometry of the reverse edge wasn't reversed.

Commit:
georef: unit tests on projection and street network path coordinates

kinnou02 commented:
Great job.
We have started to look for the replacement of this module, but it will take some time before it can be fully remove (if we can remove it)

There is almost nobody next week so we won't probably able to look at it before two weeks.

xlqian commented:
Jenkins, test this please!

kinnou02 commented on source/ed/ed_reader.cpp:
```diff
@@ -1425,10 +1425,11 @@ EdReader::get_duration (nt::Mode_e mode, float len, uint64_t source, uint64_t ta
                           "ST_LENGTH(the_geog) AS leng, e.pedestrian_allowed as pede, "
-                          "e.cycles_allowed as bike,e.cars_allowed as car from georef.edge e;";
+                          "e.cycles_allowed as bike,e.cars_allowed as car, "
+                          "ST_ASTEXT(the_geog) AS geometry from georef.edge e;";
```
It will be nice to not generate the wkt if we do not need it, on fr-nw we loose 40sec.
If your are on something else I can make a PR on your PR

kinnou02 commented on source/ed/ed2nav.cpp:
```diff
@@ -168,6 +169,10 @@ int main(int argc, char * argv[])
         ("min_non_connected_ratio,m",
          po::value<double>(&min_non_connected_graph_ratio)->default_value(0.1),
          "min ratio for the size of non connected graph")
+        ("full_street_network_geometries", po::value<bool>(&export_georef_edges_geometries)->default_value(false),
```
it's minor but you don't have to define a type, it's more a flag that we want to activate, without a type we can do this:

```
ed2nav --full_street_network_geometries <blabla>
```

then you use `vm.count("full_street_network_geometries")` for setting the value of `export_georef_edges_geometries`

kinnou02 commented on source/georef/georef.h:
```diff
@@ -82,10 +82,11 @@ struct Vertex {
 
     template<class Archive> void serialize(Archive & ar, const unsigned int) {
-        ar & way_idx & duration;
+        ar & way_idx & geom_idx & duration;
```
You need to increment `Data::data_version` in type/data.cpp
Without this an old kraken can try to load a new dataset, in most case it will lead to a segmentation fault

pbougue replied:
Done

kinnou02 commented:
Great work! I will make this configurable in tyr.

kinnou02 committed:
make configuration of full_streetnetwork_geometries accessible in tyr

kinnou02 committed:
add test in tyr

TeXitoi commented on source/ed/osm2ed.cpp:
```diff
@@ -338,6 +338,7 @@ void OSMCache::insert_edges() {
             "cycles_allowed", "cars_allowed"});
     std::stringstream geog;
     geog << std::cout.precision(10);
+    std::vector<std::string> coords;
```
you can use `navitia::type::LineString` (which is just a `typedef std::vector<GeographicalCoord`). Then, you can use boost to generate the wkt.

TeXitoi commented on source/georef/georef.cpp:
```diff
@@ -169,7 +169,16 @@ nt::GeographicalCoord Way::nearest_coord(const int number, const Graph& graph) c
+            Check if the edge way is the same that the current way.
+            Since boost::edge return only one edge between two vertices, if there is parallel edges, we aren't
+            sure it's the right one, and the geom_idx could be out of bounds.
+        */
```
If you prefer, to handle parallel edges, you can loop on `boost::out_egdes` as in nearest_edge to find the shape.

TeXitoi commented on source/georef/georef.cpp:
```diff
@@ -324,10 +333,18 @@ void ProjectionData::init(const type::GeographicalCoord & coord, const GeoRef &
+        distances[Direction::Target] = type::real_distance_from_extremity(geom, projected, true);
+    } else {
+        this->projected = coord.project(vertex1_coord, vertex2_coord).first;
+        // On calcule la distance « initiale » déjà parcourue avant d'atteindre ces extrémité d'où on effectue le calcul d'itinéraire
```
Please translate this comment in english (I know, that's not your code).

eturck commented:
Sorry I was away for the last two weeks. I will be back to work on this PR as soon as possible. 
Thank you for the feedback!

Commit:
ed: use nt::LineString and boost to generate edge WKT

Commit:
ed: don't call ST_ASTEXT if we don't want to export geometries

Commit:
ed: make --full_street_network_geometries a flag

Commit:
georef: Handle parallel edges on make_multiline

pbougue commented:
You have to rebase (but maybe in the end)

Commit:
georef: Add the Edge property to ProjectionData

It's needed if want to know exactly on which edge the projection has
been done. This is necessary to get the right edge when there is
parallel edges.

Commit:
georef: Improve path on edges with same vertex as source and target

Commit:
georef: add unit test on parallel edges / one vertex edges

Commit:
type: bump data version to 62

Commit:
georef: Use bidirectionnal for add_edge on the test find_nearest_on_same_edge

Since is_projected_on_same_edge take into account the way index we need
to add them at the same time to have the correct way idx for each edge.

eturck commented:
Hello again :).

I've made a few changes from the past feedback : 
- full_street_network_geometries is configurable in tyr, thanks to @kinnou02 ;),
- full_street_network_geometries is now a flag,
- use nt::LineString and boost in osm2ed to generate the wkt,
- don't call ST_ASTEXT if we don't want to export geometries,
- use out_edges in make_multiline to handle parallel edges,
- bump data_version to 62,

I also made some new changes. Those changes are made to handle parallel edges or edges with source and target on the same vertex : 
- In ProjectionData I've added the Edge property, this way when there is a parallel edge we know what was the exact Edge we projected on. I haven't just added the edge descriptor because it's not serializable by itself, and I didn't manage to make it serializable,
- In is_projected_on_same_edge we don't just check source and target we also check the duration and the way idx are the same,
- I made a few changes here and there to handle edges with the same vertex for source and target (dead ends ending with a roundabout with only one entrance/exit).

The last change is the biggest one. Currently, if two points are projected on the same edge we consider that the path using only this edge is the fastest way to reach those 2 points. But it's not always the case, to get from 'o' to 'd' it's faster to use v1-v2 : 

```
.------------------o----|v1
|                       |
'------------------d----|v2
```

I've changed the current way get_path works. Currently it does : 
- If we are on the same edge take a direct path,
- Otherwise use the path returned by build_path and add projection

And now I do : 
- Get the path returned by build_path and store its duration as base_duration,
- Add the projections to this path to get the total duration,
- Check if it's better to use a direct path. It is the case if : 
  - we are on the same edge
  - AND
    - the path returned by build path have a base_duration of 0 second AND it's not an edge with the same source and target (make sure we have at least an edge in the path returned by Dijkstra except if source and target are the same)
    - OR 
    - the duration on the same edge is smaller than the total duration of the path with projection

It works great when using the API but I encoutered a problem when adding a unit test. I used before "compute_path" defined in georef_test which call update_path in street_network.cpp. But when using it, I never have the path v1-v2 returned by Dijkstra.
I switched to using a StreetNetwork worker like in find_nearest_on_same_edge or the API and it's working, but only when setting a max_duration on the entry point, if I keep the default 1 second value it doesn't work.

My questions are :
- Are you ok with all those changes :)?,
- Is this a problem to have an edge property instead of an edge_t in ProjectionData? I would love to have an edge_t but I didn't find a way to make it serializable, and Google wasn't helping much...
- Is it normal to have to set a max_duration bigger than one second to have the correct path? Or is there a bug on my end?
- Is it normal that it's not working with "compute_path"? I didn't see where it was used except in the unit tests so I don't really know if that is impacting other things. What I saw is that it is not calling Djikstra the same way.

A few screenshots of what it did before : 
![parallel_before](https://cloud.githubusercontent.com/assets/5136553/18436163/26d7926c-78f7-11e6-8d92-ea7ced2ac98a.jpg)
And after : 
![parallel](https://cloud.githubusercontent.com/assets/5136553/18436181/3bdd2ef6-78f7-11e6-82dd-34b60161c8d3.jpg)

And an illustration of the use case above : 
![parralel_routing](https://cloud.githubusercontent.com/assets/5136553/18436192/4feb0fbc-78f7-11e6-9fbf-57395d3594ac.JPG)
![parallel_routing_two](https://cloud.githubusercontent.com/assets/5136553/18436194/55d63ec4-78f7-11e6-8882-31d4a228710e.JPG)

I guess we'll talk about it in a few minutes ;)

kinnou02 commented on source/tyr/tyr/binarisation.py:
```diff
@@ -457,6 +457,8 @@ def ed2nav(self, instance_config, job_id, custom_output_dir):
         if 'CITIES_DATABASE_URI' in current_app.config and current_app.config['CITIES_DATABASE_URI']:
             argv.extend(["--cities-connection-string", current_app.config['CITIES_DATABASE_URI']])
+        if instance.full_sn_geometries:
+            argv.extend(['--full_street_network_geometries', 'true'])
```
we doen't need the `true` anymore since `--full_street_network_geometries` is now a flag

Commit:
tyr: full_street_network_geometries is a flag

Don't pass a value for this flag to ed2nav

kinnou02 commented:
Seems good to me.
I think we should wait after the release to merge this PR since Artemis will probably register some changes.
